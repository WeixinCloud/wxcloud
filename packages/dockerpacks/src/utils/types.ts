export type NonEmptyArray<T> = [T, ...T[]];

export function isGenericNonEmptyArray(input: any): input is NonEmptyArray<unknown> {
  return Array.isArray(input) && isNonEmptyArray(input);
}

export function isNonEmptyArray<T = unknown>(input: Array<T>): input is NonEmptyArray<T> {
  return input.length > 0;
}
