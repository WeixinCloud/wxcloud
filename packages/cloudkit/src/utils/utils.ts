/**
 * isMatchMajorVersion('^1.2.3', '1') -> true
 *
 * support ~version & ^version
 * @param version currentVersion
 * @param majorVersion target majorVersion
 */
export const isMatchMajorVersion = (version: string, majorVersion: string) => {
  return (
    version.startsWith(`^${majorVersion}`) ||
    version.startsWith(`~${majorVersion}`) ||
    version.startsWith(majorVersion)
  );
};

/**
 * safely get deps value from packageJson
 * @param packageJson
 * @param key
 * @returns
 */
export const safeGetDepsFromPkgJSON = (packageJson: any, key: string) => {
  return packageJson?.dependencies?.[key] || packageJson?.devDependencies?.[key];
};

export const crossPlatformNpm = /^win/.test(process.platform) ? 'npm.cmd' : 'npm';
