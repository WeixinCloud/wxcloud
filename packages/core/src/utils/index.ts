export { logger } from './logger';
export { safeRequire } from './safeRequire';
export { preprocessBaseConfig } from './baseconfig';
