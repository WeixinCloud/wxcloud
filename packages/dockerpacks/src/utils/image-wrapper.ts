import { ImageTag, MatchedImage } from '@api/image';

export class ImageWrapper {
  constructor(private readonly image: MatchedImage) {}

  getFullVersion() {
    return this.image.fullVersion;
  }

  getFullVersionTag() {
    return this.image.fullVersionTag;
  }

  getMostGeneralTag() {
    if (!this.image.matchedTags.length) {
      throw new Error('unexpected matchedTags of image');
    }
    const sorted = [...this.image.matchedTags].sort(sortByGenerality);
    return sorted[0];
  }
}

function sortByGenerality(a: ImageTag, b: ImageTag) {
  switch (true) {
    case !!a.version && !!b.version:
      return countOfDots(a.version!) - countOfDots(b.version!);
    case !a.version && !b.version:
      return 0;
    case !a.version:
      return -1;
    case !b.version:
      return 1;
  }
  throw new Error('unreachable');
}

function countOfDots(input: string) {
  return [...input].filter(c => c === '.').length;
}
