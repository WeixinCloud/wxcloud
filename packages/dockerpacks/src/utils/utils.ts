export function optArg(flag: boolean, value: string | string[]) {
  return flag ? (Array.isArray(value) ? value : [value]) : [];
}
