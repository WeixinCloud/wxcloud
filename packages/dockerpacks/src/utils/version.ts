export const DOCKERPACKS_VERSION = process.env.DOCKERPACKS_VERSION || 'unknown';
