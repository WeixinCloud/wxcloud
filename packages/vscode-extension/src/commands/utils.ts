import * as vscode from 'vscode';
import ext from '../core/global';

interface ICommonCommandHandlingOptions {
  showError?: boolean;
  throwError?: boolean;
  refreshOnFinish?: boolean;
}

export function withCommonCommandHandling(
  fn: (...$args: any[]) => any,
  options: ICommonCommandHandlingOptions = {}
) {
  const { showError = true, throwError = true, refreshOnFinish = true } = options;

  return async (...args: any[]) => {
    try {
      return await fn(...args);
    } catch (e) {
      showError && vscode.window.showErrorMessage(`${e}`);
      if (throwError) {
        throw e;
      }
    } finally {
      refreshOnFinish && ext.wxContainersProvider.refresh();
    }
  };
}
