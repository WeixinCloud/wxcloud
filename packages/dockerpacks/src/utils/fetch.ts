import { merge } from 'lodash';
import fetch, { RequestInfo, RequestInit } from 'node-fetch';
import { DOCKERPACKS_VERSION } from './version';

export function fetchUrl(url: RequestInfo, init?: RequestInit) {
  const source: RequestInit = {
    headers: { 'User-Agent': `Dockerpacks ${DOCKERPACKS_VERSION}` }
  };
  return fetch(url, merge(init ?? {}, source));
}
