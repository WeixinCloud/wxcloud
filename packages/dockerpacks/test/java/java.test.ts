import { describe } from 'vitest';
import { BuilderTestCase, runTest } from '@test/utils';
import path from 'path';

const CASES: BuilderTestCase[] = [
  {
    id: 'with-jar',
    promptAnswers: {
      environments: []
    }
  },
  {
    id: 'wxcloudrun-sprintboot',
    promptAnswers: {
      environments: [],
      entrypointJar: 'dist/app.jar'
    }
  },
  {
    id: 'starters-kotlin-spring',
    promptAnswers: {
      environments: [],
      entrypointJar: 'dist/app.jar'
    }
  }
];

const FIXTURES_PATH = path.join(__dirname, 'fixtures');

describe('builders test (java cases)', () => {
  describe.each(CASES)('case $id', testCase => {
    runTest(FIXTURES_PATH, testCase);
  });
});
