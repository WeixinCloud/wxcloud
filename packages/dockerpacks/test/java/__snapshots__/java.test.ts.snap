// Vitest Snapshot v1

exports[`builders test (java cases) > case starters-kotlin-spring > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用 Gradle 官方镜像
FROM gradle:6-jdk11

# 设置容器内的当前目录
WORKDIR /app

# 将包括源文件在内的所有文件拷贝到容器中
COPY . .

# 编译项目
RUN gradle clean assemble -x test --quiet

# 服务暴露的端口
EXPOSE 8080

# 运行项目
CMD [\\"java\\", \\"-jar\\", \\"build/libs/demo-0.0.1-SNAPSHOT.jar\\"]",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md",
  },
}
`;

exports[`builders test (java cases) > case with-jar > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用 Maven 官方镜像
FROM maven:3-jdk-11

# 设置容器内的当前目录
WORKDIR /app

# 将包括源文件在内的所有文件拷贝到容器中
COPY . .

# 编译项目
RUN mvn clean package --batch-mode -DskipTests -Dhttp.keepAlive=false --quiet

# 服务暴露的端口
EXPOSE 3000

# 运行项目
CMD [\\"java\\", \\"-jar\\", \\"target/test.jar\\"]",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md",
  },
}
`;

exports[`builders test (java cases) > case wxcloudrun-sprintboot > should have correct outputs 1`] = `
{
  "dockerfile": "# 由 Dockerpacks 自动生成
# 本 Dockerfile 可能不能完全覆盖您的项目需求，若遇到问题请根据实际情况修改或询问客服

# 使用 Maven 官方镜像
FROM maven:3-jdk-11

# 设置容器内的当前目录
WORKDIR /app

# 将包括源文件在内的所有文件拷贝到容器中
COPY . .

# 编译项目
RUN mvn -s settings.xml clean package --batch-mode -DskipTests -Dhttp.keepAlive=false --quiet

# 服务暴露的端口
EXPOSE 80

# 运行项目
CMD [\\"java\\", \\"-jar\\", \\"springboot-wxcloudrun-1.0.jar\\"]",
  "files": Map {
    ".dockerignore" => ".git
.gitignore
.dockerignore
Dockerfile*
LICENSE
*.md",
  },
}
`;
